File: benchmark/languages/go/work/day04/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func readGrid(filename string) [][]byte {
	f, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	defer f.Close()

	var grid [][]byte
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			continue
		}
		grid = append(grid, []byte(line))
	}
	return grid
}

// countNeighbors counts how many of the 8 adjacent cells are '@'
func countNeighbors(grid [][]byte, r, c int) int {
	rows := len(grid)
	cols := len(grid[0])
	count := 0
	for dr := -1; dr <= 1; dr++ {
		for dc := -1; dc <= 1; dc++ {
			if dr == 0 && dc == 0 {
				continue
			}
			nr, nc := r+dr, c+dc
			if nr >= 0 && nr < rows && nc >= 0 && nc < cols {
				if grid[nr][nc] == '@' {
					count++
				}
			}
		}
	}
	return count
}

// accessible returns true if a roll at (r,c) has fewer than 4 '@' neighbors
func accessible(grid [][]byte, r, c int) bool {
	return grid[r][c] == '@' && countNeighbors(grid, r, c) < 4
}

func part1(grid [][]byte) int {
	count := 0
	for r := range grid {
		for c := range grid[r] {
			if accessible(grid, r, c) {
				count++
			}
		}
	}
	return count
}

func copyGrid(grid [][]byte) [][]byte {
	g := make([][]byte, len(grid))
	for i, row := range grid {
		g[i] = make([]byte, len(row))
		copy(g[i], row)
	}
	return g
}

func part2(grid [][]byte) int {
	g := copyGrid(grid)
	total := 0

	for {
		// Find all accessible rolls in current state
		var toRemove [][2]int
		for r := range g {
			for c := range g[r] {
				if accessible(g, r, c) {
					toRemove = append(toRemove, [2]int{r, c})
				}
			}
		}
		if len(toRemove) == 0 {
			break
		}
		// Remove them all at once
		for _, pos := range toRemove {
			g[pos[0]][pos[1]] = '.'
		}
		total += len(toRemove)
	}
	return total
}

func main() {
	part := 0
	if len(os.Args) > 1 {
		fmt.Sscanf(os.Args[1], "%d", &part)
	}

	grid := readGrid("input.txt")

	if part == 1 || part == 0 {
		fmt.Printf("Part 1: %d\n", part1(grid))
	}
	if part == 2 || part == 0 {
		fmt.Printf("Part 2: %d\n", part2(grid))
	}
}

File: benchmark/languages/c/work/day04/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_ROWS 1024
#define MAX_COLS 1024

static char grid[MAX_ROWS][MAX_COLS];
static int rows, cols;

static int dx[] = {-1,-1,-1, 0, 0, 1, 1, 1};
static int dy[] = {-1, 0, 1,-1, 1,-1, 0, 1};

static int count_neighbors(int r, int c) {
    int cnt = 0;
    for (int d = 0; d < 8; d++) {
        int nr = r + dx[d];
        int nc = c + dy[d];
        if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && grid[nr][nc] == '@')
            cnt++;
    }
    return cnt;
}

static int is_accessible(int r, int c) {
    return grid[r][c] == '@' && count_neighbors(r, c) < 4;
}

int main(int argc, char *argv[]) {
    int part = 0;
    if (argc >= 2) part = atoi(argv[1]);

    FILE *f = fopen("input.txt", "r");
    if (!f) { fprintf(stderr, "Cannot open input.txt\n"); return 1; }

    rows = 0;
    cols = 0;
    char line[MAX_COLS + 10];
    while (fgets(line, sizeof(line), f)) {
        int len = (int)strlen(line);
        while (len > 0 && (line[len-1] == '\n' || line[len-1] == '\r'))
            len--;
        if (len == 0) continue;
        if (cols == 0) cols = len;
        memcpy(grid[rows], line, len);
        grid[rows][len] = '\0';
        rows++;
        if (rows >= MAX_ROWS) break;
    }
    fclose(f);

    // Part 1: count accessible rolls
    long long part1 = 0;
    for (int r = 0; r < rows; r++)
        for (int c = 0; c < cols; c++)
            if (is_accessible(r, c))
                part1++;

    // Part 2: repeatedly remove accessible rolls until none left
    // Use a queue-based BFS approach
    // We need a copy of grid for part2
    static char grid2[MAX_ROWS][MAX_COLS];
    for (int r = 0; r < rows; r++)
        memcpy(grid2[r], grid[r], cols + 1);

    // Use a boolean "in_queue" array and a queue
    // Queue stores (r,c) encoded as r*MAX_COLS+c
    static int in_queue[MAX_ROWS * MAX_COLS];
    int *queue = malloc(MAX_ROWS * MAX_COLS * sizeof(int));
    if (!queue) { fprintf(stderr, "OOM\n"); return 1; }

    memset(in_queue, 0, sizeof(in_queue));

    int qhead = 0, qtail = 0;

    // Helper: count neighbors in grid2
    // We'll inline this as a lambda-style check

    // Initialize queue with all accessible cells
    for (int r = 0; r < rows; r++) {
        for (int c = 0; c < cols; c++) {
            if (grid2[r][c] == '@') {
                // count neighbors in grid2
                int cnt = 0;
                for (int d = 0; d < 8; d++) {
                    int nr = r + dx[d], nc = c + dy[d];
                    if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && grid2[nr][nc] == '@')
                        cnt++;
                }
                if (cnt < 4) {
                    int idx = r * MAX_COLS + c;
                    if (!in_queue[idx]) {
                        in_queue[idx] = 1;
                        queue[qtail++] = idx;
                    }
                }
            }
        }
    }

    long long part2 = 0;
    while (qhead < qtail) {
        int idx = queue[qhead++];
        int r = idx / MAX_COLS;
        int c = idx % MAX_COLS;
        in_queue[idx] = 0;

        // Check if still accessible in grid2
        if (grid2[r][c] != '@') continue;
        int cnt = 0;
        for (int d = 0; d < 8; d++) {
            int nr = r + dx[d], nc = c + dy[d];
            if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && grid2[nr][nc] == '@')
                cnt++;
        }
        if (cnt >= 4) continue; // not accessible anymore (shouldn't happen since we enqueue when accessible)

        // Remove this roll
        grid2[r][c] = '.';
        part2++;

        // Check neighbors: they might now be accessible
        for (int d = 0; d < 8; d++) {
            int nr = r + dx[d], nc = c + dy[d];
            if (nr < 0 || nr >= rows || nc < 0 || nc >= cols) continue;
            if (grid2[nr][nc] != '@') continue;
            // Count its neighbors
            int ncnt = 0;
            for (int d2 = 0; d2 < 8; d2++) {
                int nnr = nr + dx[d2], nnc = nc + dy[d2];
                if (nnr >= 0 && nnr < rows && nnc >= 0 && nnc < cols && grid2[nnr][nnc] == '@')
                    ncnt++;
            }
            if (ncnt < 4) {
                int nidx = nr * MAX_COLS + nc;
                if (!in_queue[nidx]) {
                    in_queue[nidx] = 1;
                    queue[qtail++] = nidx;
                }
            }
        }
    }

    free(queue);

    if (part == 1) {
        printf("Part 1: %lld\n", part1);
    } else if (part == 2) {
        printf("Part 2: %lld\n", part2);
    } else {
        printf("Part 1: %lld\n", part1);
        printf("Part 2: %lld\n", part2);
    }

    return 0;
}
